butte-build/src/ir/types.rs

```diff
@@ -0,0 +1,518 @@
+        match self {
+            // Apparently there are fixed sized arrays
+            // which are considered scalar if their contents
+            // is scalar
+
+            // However they're not currently supported by butte's parser
+            Type::String | Type::Array(..) => false,
+            Type::Custom(CustomTypeRef { ty, .. }) => match ty {
+                // Apparently only C++ impl supports union in structs
+                // TODO check if unions containing only structs are considered scalar?
+                CustomType::Table | CustomType::Union { .. } => false,
+                CustomType::Struct { .. } | CustomType::Enum { .. } => true,
+            },
+            _ => true,
+        }
+    }
+
+    pub fn is_union(&self) -> bool {
+        match self {
+            Type::Custom(CustomTypeRef { ty, .. }) => match ty {
+                CustomType::Union { .. } => true,
+                _ => false,
+            },
+            _ => false,
+        }
+    }
+
+    pub fn is_enum(&self) -> bool {
+        match self {
+            Type::Custom(CustomTypeRef { ty, .. }) => match ty {
+                CustomType::Enum { .. } => true,
+                _ => false,
+            },
+            _ => false,
+        }
+    }
+
+    // For unions, we generate an enum type "companion" that's just a simple
+    // ubyte based enum with all the union variants, + None (default 0)
+    //
+    // This method lets us generate the companion type name from the union.
+    pub fn make_union_enum_type_companion(&self) -> Option<&DottedIdent<'a>> {
+        match self {
+            Type::Custom(CustomTypeRef { ty, .. }) => match ty {
+                CustomType::Union { ref enum_ident, .. } => Some(enum_ident),
+                _ => None,
+            },
+            _ => None,
+        }
+    }
+}
+
+impl Display for Type<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Type::Bool => write!(f, "bool"),
+            Type::Byte => write!(f, "i8"),
+            Type::UByte => write!(f, "u8"),
+            Type::Short => write!(f, "i16"),
+            Type::UShort => write!(f, "u16"),
+            Type::Int => write!(f, "i32"),
+            Type::UInt => write!(f, "u32"),
+            Type::Float => write!(f, "f32"),
+            Type::Long => write!(f, "i64"),
+            Type::ULong => write!(f, "u64"),
+            Type::Double => write!(f, "f64"),
+            Type::Int8 => write!(f, "i8"),
+            Type::UInt8 => write!(f, "u8"),
+            Type::Int16 => write!(f, "i16"),
+            Type::UInt16 => write!(f, "u16"),
+            Type::Int32 => write!(f, "i32"),
+            Type::UInt32 => write!(f, "u32"),
+            Type::Int64 => write!(f, "i64"),
+            Type::UInt64 => write!(f, "u64"),
+            Type::Float32 => write!(f, "f32"),
+            Type::Float64 => write!(f, "f64"),
+            Type::String => write!(f, "&str"),
```

Comment:
Should this be `&'static str`?

Reply:
In practice this doesn't matter much as this is just used for "displaying", e.g in doc we generate (though, we never generate doc using this, as strings can't have defaults).  In general though, the string slice returned from buffers will usually have the (elided) lifetime of the buffer, and that lifetime wouldn't be `'static`.

Reply:
Sounds good